src/core/lib/Mime.mjs

```diff
@@ -0,0 +1,310 @@
+                        outObj.fields[obj[0]] = Mime._extractField(mimePart, obj[1]);
+                        break;
+                    case 3:
+                        outObj.fields[obj[0]] = Mime._extractField(mimePart, obj[1], obj[2]);
+                        break;
+                    default:
+                        throw new OperationError("Invalid extraction in headers. Invalid Array size.");
+                }
+            });
+            output.push(outObj);
+        }, recursive);
+        return output;
+    }
+
+    /**
+     * Common helper function to decode Mime encoded words in headers.
+     *
+     * @param {boolean} recursive
+     */
+    decodeHeaderWords(recursive=true) {
+        Mime.walkMime(this.mimeObj, function(mimePart) {
+            if (mimePart.rawHeader) {
+                mimePart.rawHeader = Mime.replaceEncodedWord(mimePart.rawHeader);
+            }
+        }, recursive);
+    }
+
+    /**
+     * Common helper function to decode Mime bodies.
+     *
+     * @param {boolean} recursive
+     */
+    decodeMimeObjects(recursive=true) {
+        Mime.walkMime(this.mimeObj, function(mimePart) {
+            Mime.decodeMimeMessage(mimePart);
+        }, recursive);
+    }
+
+    /**
+     * Walks a MIME document and returns a Mime Object.
+     *
+     * @param {string} mimeData
+     * @returns {object}
+     */
+    static _parseMime(mimeData) {
+        const mimeObj = Mime._splitParseHead(mimeData);
+        const contType = Mime._extractField(mimeObj, "content-type");
+        const boundary = Mime._extractField(mimeObj, "content-type", "boundary");
+        if (mimeObj.body && contType && contType.startsWith("multipart/")) {
+            if (!boundary) {
+                throw new OperationError("Invalid mulitpart section no boundary");
+            }
+            const sections = [];
+            for (const val of Mime._splitMultipart(mimeObj.body, boundary)) {
+                sections.push(Mime._parseMime(val));
+            }
+            if (sections.length) {
+                mimeObj.body = sections;
+            }
+        }
+        return mimeObj;
+    }
+
+    /**
+     * Executes a function on a mime object. These methods should modify the mimeObj.
+     *
+     * @param {Object} mimeObj
+     * @param {function} methods
+     * @param {boolean} recursive
+     */
+    static walkMime(mimeObj, method, recursive=true) {
+        const contType = Mime._extractField(mimeObj, "content-type");
+        method(mimeObj);
+        if (recursive && mimeObj.body && Array.isArray(mimeObj.body) && contType && contType.startsWith("multipart/")) {
+            mimeObj.body.forEach(function(obj) {
+                Mime.walkMime(obj, method);
+            });
+        }
+    }
+
+    /**
+     * Attempts to decode a mimeObj's data by applying appropriate character and content decoders based on the header data.
+     *
+     * @param {Object} mimeObj
+     */
+    static decodeMimeMessage(mimeObj) {
```

Comment:
In the description here you say 'Attempts to..' - what is the consequence of it failing to decode? Does it just leave the `mimeObj`'s data encoded?

Reply:
Oh I see the `_decodeMimeData` throws. Please can you add a `@throws` statement to the function comment? Its helpful for reviewers like me now and in the future.